core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java

```diff
@@ -68,38 +72,46 @@ public static ParseItem get(final ParseGraph graph, final Token definition) {
+        if (graphList.isEmpty()) { return complete(() -> values); }
+        final ParseGraph graph = graphList.head;
+        if (graph.isEmpty()) { return intermediate(() -> getAllValuesRecursive(graphList.tail, values, definition)); }
         final ParseItem head = graph.head;
         if (head.isValue() && head.asValue().definition.equals(definition)) {
-            return tailResults.add(head.asValue());
+            return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail), values.add(head.asValue()), definition));
         }
         if (head.isGraph()) {
-            return tailResults.add(getAllValuesRecursive(head.asGraph(), definition));
+            return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail).add(graph.head.asGraph()), values, definition));
         }
-        return tailResults;
+        return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail), values, definition));
     }
 
     public static ImmutableList<ParseItem> getAllRoots(final ParseGraph graph, final Token definition) {
-        checkNotNull(graph, "graph");
-        checkNotNull(definition, "definition");
-        return getAllRootsRecursive(graph, null, definition);
+        return getAllRootsRecursive(ImmutableList.create(new Pair(checkNotNull(graph, "graph"), null)), checkNotNull(definition, "definition"), new ImmutableList<>()).computeResult();
     }
 
-    private static ImmutableList<ParseItem> getAllRootsRecursive(final ParseItem item, final ParseGraph parent, final Token definition) {
-        final ImmutableList<ParseItem> result = item.getDefinition().equals(definition) && (parent == null || !parent.getDefinition().equals(definition))
-            ? ImmutableList.create(item)
-            : new ImmutableList<>();
+    private static SafeTrampoline<ImmutableList<ParseItem>> getAllRootsRecursive(final ImmutableList<Pair> list, final Token definition, final ImmutableList<ParseItem> result) {
+        if (list.isEmpty()) { return complete(() -> result); }
+        final ParseItem item = list.head.item;
+        final ParseGraph parent = list.head.parent;
+        final ImmutableList<ParseItem> nextResult = item.getDefinition().equals(definition) && (parent == null || !parent.getDefinition().equals(definition)) ? result.add(item) : result;
         if (item.isGraph() && !item.asGraph().isEmpty()) {
-            return result
-                .add(getAllRootsRecursive(item.asGraph().tail, item.asGraph(), definition))
-                .add(getAllRootsRecursive(item.asGraph().head, item.asGraph(), definition));
+            return intermediate(() -> getAllRootsRecursive(list.tail.add(new Pair(item.asGraph().head, item.asGraph())).add(new Pair(item.asGraph().tail, item.asGraph())), definition, nextResult));
+        }
+        return intermediate(() -> getAllRootsRecursive(list.tail, definition, nextResult));
+    }
+
+    static class Pair {
```

Comment:
Makes sense, but I don't want to clutter it too much with helper method definitions.